v2alpha1/pkg/stub/stub.go

```diff
@@ -0,0 +1,724 @@
+		return nil, err
+	}
+
+	if err := stub.setupHandlers(); err != nil {
+		return nil, errors.New("internal error: invalid plugin, can't handle any NRI events")
+	}
+
+	for _, o := range opts {
+		if err = o(stub); err != nil {
+			return nil, err
+		}
+	}
+
+	if stub.name == "" {
+		if err := stub.identityFromBinary(); err != nil {
+			return nil, err
+		}
+	}
+
+	// connect to NRI Runtime Service socket and set up multiplexing
+	if stub.conn == nil {
+		if stub.conn, err = Connect(); err != nil {
+			return nil, err
+		}
+	}
+	stub.rpcm = multiplex.Multiplex(stub.conn)
+
+	// create NRI Plugin Service ttrpc server
+	if stub.rpcl, err = stub.rpcm.Listen(multiplex.PluginServiceConn); err != nil {
+		return nil, err
+	}
+	if stub.rpcs, err = ttrpc.NewServer(); err != nil {
+		return nil, errors.Wrap(err, "failed to create ttrpc server")
+	}
+	api.RegisterPluginService(stub.rpcs, stub)
+
+	// create NRI Runtime Service ttrpc client
+	conn, err := stub.rpcm.Open(multiplex.RuntimeServiceConn)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to multiplex ttrpc client connection")
+	}
+	stub.rpcc = ttrpc.NewClient(conn,
+		ttrpc.WithOnClose(func() {
+			stub.connClosed()
+		}),
+	)
+	stub.runtime = api.NewRuntimeClient(stub.rpcc)
+
+	log.Infof(noCtx, "Created plugin %s (%s, handles %s)", stub.Name(),
+		filepath.Base(os.Args[0]), stub.events.PrettyString())
+
+	return stub, nil
+}
+
+// Run the plugin (start event processing, wait for an error or getting stopped).
+func (stub *stub) Run(ctx context.Context) error {
+	var err error
+
+	if err = stub.Start(ctx); err != nil {
+		return err
+	}
+
+	select {
+	case err = <-stub.errC:
+		return err
+	case <-stub.stopC:
+	}
+
+	return nil
+}
+
+// Start processing events and requests relayed by the NRI Runtime.
+func (stub *stub) Start(ctx context.Context) error {
+	var err error
+
+	stub.errC = make(chan error, 1)
+	go func() {
+		err := stub.rpcs.Serve(ctx, stub.rpcl)
+		stub.errC <- err
+	}()
+
+	stub.cfgC = make(chan error, 1)
+	if err := stub.register(ctx); err != nil {
+		stub.Stop()
+		return err
+	}
+	err = <-stub.cfgC
+
+	if err == nil {
+		log.Infof(ctx, "Started plugin %s...", stub.Name())
+	}
+
+	return err
+}
+
+// Stop the plugin.
+func (stub *stub) Stop() error {
+	var err error
+
+	log.Infof(noCtx, "Stopping plugin %s...", stub.Name())
+
+	stub.stopOnce.Do(func() {
+		stub.rpcm.Close()
+		stub.rpcl.Close()
+		stub.rpcs.Close()
+		stub.rpcc.Close()
+		close(stub.errC)
+		close(stub.stopC)
+		if stub.cfgC != nil {
+			close(stub.cfgC)
+		}
+	})
+	return err
+}
+
+// Wait for the plugin to stop.
+func (stub *stub) Wait() {
+	<-stub.stopC
+}
+
+// Name returns the full indexed name of the plugin.
+func (stub *stub) Name() string {
+	return stub.idx + "-" + stub.name
+}
+
+// Connect returns a connection to the NRI/runtime.
+func Connect() (stdnet.Conn, error) {
+	conn, err := ConnFromEnv()
+	if err == ErrNoConn {
+		conn, err = Dial(api.DefaultSocketPath)
```

Comment:
Do you think that `Connect()` could use an environment variable to override `api.DefaultSocketPath`? Or maybe reuse `NRI_PLUGIN_SOCKET` with two different schemes. For instance:
```
NRI_PLUGIN_SOCKET=fd://42
NRI_PLUGIN_SOCKET=unix:///var/run/nri-test.sock
```